### PR TITLE
Anable the use of generators to reduce memory usage via a LazyMultiSet

### PR DESCRIPTION
**Pull request by samwillis**
Also converts almost all operators to use them wherever possible.

We should benchmark this against main to see how what it does with the memory usage and allocations.

### CODE REVIEW

File: packages/d2mini/src/indexes.ts [hunk 1]
```diff
@@ -124,4 +124,32 @@ export class Index<K, V> {
 
     return new MultiSet(result)
   }
+
+  *lazyJoin<V2>(other: Index<K, V2>): Generator<[[K, [V, V2]], number], void, unknown> {
```
Comment (kevin-dp) [inline packages/d2mini/src/indexes.ts:128]: Let's remove the duplication between both branches like i did here: https://github.com/electric-sql/d2ts/pull/58/commits/ac24ff08e2bea46dab3fbc00e90b59f3f95d25a1

File: packages/d2mini/src/operators/topK.ts [hunk 2]
```diff
@@ -34,9 +34,8 @@ export function topK<
     const reduced = stream.pipe(
       reduce((values) => {
         // `values` is a list of tuples, first element is the value, second is the multiplicity
-        const consolidated = new MultiSet(values).consolidate()
-        const sortedValues = consolidated
-          .getInner()
+        const consolidated = LazyMultiSet.fromArray(values).consolidate()
```
Comment (kevin-dp) [inline packages/d2mini/src/operators/topK.ts:37]: Why creating a lazy multiset to then consolidate it (which materializes it) to then immediately copy it into an array. Seems like a lot of overhead for something we could do by reducing once over the original array (i.e. loop over `values` and consolidate it into an array of consolidated values).

File: packages/d2mini/tests/multiset.test.ts [hunk 3]
```diff
@@ -108,3 +108,101 @@ describe('MultiSet', () => {
...
+    it('should negate multiplicities', () => {
+      const negated = a.negate()
+      expect(negated.getInner()).toEqual([
+        [['apple', '$5'], -2],
+        [['banana', '$2'], -1],
+      ])
+    })
+  })
+
+  it('should consolidate correctly', () => {
```
Comment (kevin-dp) [inline packages/d2mini/tests/multiset.test.ts:173]: Let's also check the corner case where the multiplicities for a certain value cancel out and that the result does not contain the value.

**Review by kevin-dp — APPROVED** [review]
LGTM implementation-wise. However, i do think we should carefully benchmark performance and memory usage before merging as it does complicate the code quite a bit so i want to be 100% sure this is the way to go.
